atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java

```diff
@@ -1313,4 +1329,34 @@ private static Integer getMaxRowsFromBatchHint(@Nullable Integer batchHint) {
     private interface ReadWriteTask<T> {
         T run(DbReadTable readTable, DbWriteTable writeTable);
     }
+
+    static class ResultWithToken {
+        SetMultimap<Cell, Long> results;
+        byte[] currentRow = null;
+        boolean finishedRange = false;
+
+        ResultWithToken() {
+            results = HashMultimap.create();
+        }
+
+        boolean isEmpty() {
+            return results.isEmpty();
+        }
+
+        int size() {
+            return results.size();
+        }
+
+        void storeRow(SetMultimap<Cell, Long> rowBuffer) {
+            results.putAll(rowBuffer);
+        }
+
+        void storePreviousRowIfComplete(AgnosticLightResultRow cellResult, SetMultimap<Cell, Long> rowBuffer) {
+            if (!Arrays.equals(currentRow, cellResult.getBlob("row_name"))) {
```

Comment:
Should also pull the row name out to a local to avoid reading twice:

```
byte[] rowName = cellResult.getBytes("row_name");
if (!Arrays.equals(currentRow, rowName)) {
    currentRow = rowName;
```